ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,416 @@
+
+    // Block all possible signals to avoid receiving some in the child.
+    // Signal mask is inherited to new process/thread will start with
+    // all signals disabled and we can safely change them.
+    sigfillset(&sigset);
+    pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
+
+    // fork
+    err = 0;
+    res->pid = vfork();
+    if (res->pid < 0) {
+        err = errno;
+    } else if (res->pid == 0) {
+        // child
+        if (pipe_fds[0] >= 0)
+            close(pipe_fds[0]);
+        execve(args[0], args, envs);
+        // keep compatibility with forkexecd daemon.
+        _exit(errno == ENOENT ? 127 : 126);
+    }
+
+    // Restore thread state
+    pthread_sigmask(SIG_SETMASK, &old_sigset, NULL);
+    pthread_setcancelstate(cancellation_state, NULL);
+
+    // We don't need writing pipe anymore and we need to detect
+    // if closed so we can't keep it open
+    close_fd(&pipe_fds[1]);
+
+    if (err != 0) {
+        close_fd(&pipe_fds[0]);
+        res->err_msg = "vfork";
+        return err;
+    }
+
+    // Handle errors from helper
+    if (pipe_fds[0] >= 0) {
+        int readed;
+        // Note that buffer is small and written atomically by
+        // the helper, no reason for the kernel to split it.
+        while ((readed = read(pipe_fds[0], &res->msg, sizeof(res->msg))) < 0
+               && errno == EINTR)
+            continue;
+        close_fd(&pipe_fds[0]);
+        if (readed != 0 && readed < offsetof(msg_t, msg_buf) + 1) {
+            // This should never happen !!!
+            // At this point the process is created and we have a pid so
+            // we cannot just return an error.
+            // We could try to wait the process but it should fail, let
+            // returns success and let caller read process status result.
+            return 0;
+        }
+        res->msg.msg_buf[sizeof(res->msg.msg_buf) - 1] = 0;
+        if (readed > 0) {
+            // Wait the process otherwise we'll have a zombie
+            reap_pid(res->pid);
+
+            res->err_msg = res->msg.msg_buf;
+            return res->msg.err;
+        }
+    }
+    return 0;
+}
+
+CAMLprim value
+caml_safe_exec_with_helper(value args, value environment)
+{
+    CAMLparam2(args, environment);
+
+    // Copy parameters to C
+    char **c_args = copy_string_list(args);
+    char **c_envs = copy_string_list(environment);
+    if (!c_envs || !c_args) {
+        free(c_envs);
+        free(c_args);
+        caml_raise_out_of_memory();
+    }
+
+    // potentially slow section, release Ocaml engine
+    caml_enter_blocking_section();
+
+    safe_exec_result res;
+    int err = safe_exec_with_helper(&res, c_args, c_envs);
+
+    free(c_envs);
+    free(c_args);
+
+    caml_leave_blocking_section();
+
+    // error, notify with an exception
+    if (err != 0)
+        unix_error(err, res.err_msg, Nothing);
+
+    CAMLreturn(Val_int(res.pid));
+}
+
+CAMLprim value
+caml_pidwaiter_dontwait(value pid_val)
```

Comment:
Why do you need to create a thread to do this? Won't the double fork in the vfork_helper mean that the pid gets reparented to pid 1 so systemd would be responsible for reaping the process?

Reply:
The double fork is done only for syslog handling and the child is the logger, not the launched program